### PR TITLE
Use CMake package in data-sealing sample

### LINKED ISSUE #300: Users of Open Enclave repo have a sample illustrating integrating their app CMAKE project with Open Enclave's CMAKE config
In relation to making the Open Enclave project easy to ingest via cmake directly (#204), we should consider including a sample that illustrates and tests that we have not broken that experience. This is a  narrow scenario that is a much lower priority to illustrate via samples, and currently is for an internal audience only. 

This is a separate task from having cmake samples designed specifically to illustrate converged Windows & Linux enclave apps; that story is tracked under #378.

### PR DESCRIPTION
**Pull request by andschwa**
This introduces a sample build using the CMake package for arguably our most complicated sample, `data-sealing`. It also introduces a CMake script to configure, build, and test the sample from the installed location, which will replace `test-samples.sh` when builds are written for the remaining samples (see #1357).

```
...
Host: operation succeeded: completed as expected
Host: Terminating enclaves
Host: enclave terminated.
Host: enclave terminated.
Host: enclave terminated.
Host: Sample completed successfully.
```

Resolves #300, #748, #887, #1538 and #1539.

### CODE REVIEW

File: samples/data-sealing/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+project("Data Sealing Sample" LANGUAGES C CXX)
+
+find_package(Threads REQUIRED)
+
+find_package(OpenEnclave CONFIG REQUIRED)
+add_executable(openenclave::oeedger8r IMPORTED)
+set_target_properties(openenclave::oeedger8r PROPERTIES IMPORTED_LOCATION ${OE_BINDIR}/oeedger8r)
+
+add_custom_command(OUTPUT ${CMAKE_SOURCE_DIR}/common/datasealing_t.c
```
Comment (CodeMonkeyLeet) [inline samples/data-sealing/CMakeLists.txt]: Why not just break this across the subdirectories, specifically with `common` having its own CMakeLists.txt? Could this just be a custom target that the enclave subdirectories depend on? It seems like something the Makefile implementation should have done as well, instead of having it as part of `enc1` specifically.

File: samples/data-sealing/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+# Enclave 1
+add_executable(enclave_a_v1
+  enc1/ecalls.cpp
+  common/dispatcher.cpp
+  common/keys.cpp
+  common/datasealing_t.c)
+
+target_link_libraries(enclave_a_v1 openenclave::oeenclave openenclave::oelibcxx)
+
+add_custom_command(OUTPUT ${CMAKE_SOURCE_DIR}/enc1/private.pem ${CMAKE_SOURCE_DIR}/enc1/public.pem
```
Comment (CodeMonkeyLeet) [inline samples/data-sealing/CMakeLists.txt]: Similarly for the generation of the private keys, it's probably better to have `private_a.pem` and `private_b.pem` generated as part of `common`, and then have the enclave builds depend on those.

I would not remove the Makefiles build, but they should be updated to be consistent with the CMake one.

File: samples/data-sealing/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+add_custom_command(OUTPUT enclave_b.signed
+  DEPENDS ${CMAKE_SOURCE_DIR}/enc3/private.pem
+  COMMAND openenclave::oesign sign $<TARGET_FILE:enclave_b> ${CMAKE_SOURCE_DIR}/enc3/data-sealing.conf ${CMAKE_SOURCE_DIR}/enc3/private.pem
+  WORKING_DIRECTORY ${CMAKE_BINARY_DIR})
+
+# Host
+add_subdirectory(host)
+
+add_custom_target(run
```
Comment (CodeMonkeyLeet) [inline samples/data-sealing/CMakeLists.txt]: For my understanding, is this a common pattern in CMake? I assume it was done to mirror the Makefiles, but if there's a more idiomatic way to do this in CMake (e.g. ctest?) then we should use it as it would be more intuitive for its target audience.

File: samples/data-sealing/CMakeLists.txt (hunk 1 above)
Comment (andschwa) [reply samples/data-sealing/CMakeLists.txt]: See my description and commit for exactly why this happened this way 😉

File: samples/data-sealing/CMakeLists.txt (hunk 2 above)
Comment (andschwa) [reply samples/data-sealing/CMakeLists.txt]: Again, I agree, but this was the implementation _without_ changing the existing build or project structure.

File: samples/data-sealing/CMakeLists.txt (hunk 3 above)
Comment (andschwa) [reply samples/data-sealing/CMakeLists.txt]: So much more idiomatic ways, but this pass was to mirror GNU Make 😋 If we can remove the Makefiles, I can write this way better. My proposal to @soccerGB was to have two sets of samples, `samples/pkg-config` and `samples/cmake` so we don't have to do weird things.

File: enclave/core/CMakeLists.txt [hunk 4]
```diff
@@ -132,7 +132,10 @@ target_compile_options(oecore INTERFACE $<$<COMPILE_LANGUAGE:CXX>:-nostdinc++>)
 target_compile_definitions(oecore
     PUBLIC
     OE_BUILD_ENCLAVE
-    OE_API_VERSION=2)
+    # NOTE: This definition is public to the rest of our project's
```
Comment (andschwa) [inline enclave/core/CMakeLists.txt]: This is in #1513 but required to build this right now. Waiting for CI to start working again.

File: host/CMakeLists.txt [hunk 5]
```diff
@@ -154,7 +154,11 @@ endif ()
 
 # Compile definitions and options
 target_compile_definitions(oehost
-  PUBLIC -DOE_API_VERSION=2
+  PUBLIC
+  # NOTE: This definition is public to the rest of our project's
```
Comment (andschwa) [inline host/CMakeLists.txt]: Ditto.

File: samples/data-sealing/common/shared.h [hunk 6]
```diff
@@ -4,6 +4,8 @@
 #ifndef _SHARED_H
 #define _SHARED_H
 
+#include <stddef.h>
```
Comment (andschwa) [inline samples/data-sealing/common/shared.h]: This was required since we use a `size_t` below, not sure how it built without it previously.

File: samples/data-sealing/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,49 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+cmake_minimum_required(VERSION 3.10)
+
+project("Data Sealing Sample" LANGUAGES C CXX)
+
+find_package(Threads REQUIRED)
```
Comment (andschwa) [inline samples/data-sealing/CMakeLists.txt]: We should document that this is required. (In PR checklist.)

File: samples/data-sealing/CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+
+find_package(OpenEnclave CONFIG REQUIRED)
+
+# TODO: Remove this when the package exports the target.
+add_executable(openenclave::oeedger8r IMPORTED)
```
Comment (andschwa) [inline samples/data-sealing/CMakeLists.txt]: I've heard on IRC that we can move this into our CMake package by manually injecting some stuff into it, which would be nice. It's also in the PR checklist, but I think we can do it later.

File: samples/data-sealing/common/CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,12 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+# Use the edger8r to generate C bindings from the EDL file.
+add_custom_command(OUTPUT datasealing_t.h datasealing_t.c
+  COMMAND openenclave::oeedger8r --trusted ${CMAKE_SOURCE_DIR}/datasealing.edl)
+
+# Create a library common to each of our three enclaves.
+add_library(common OBJECT dispatcher.cpp keys.cpp ${CMAKE_CURRENT_BINARY_DIR}/datasealing_t.c)
```
Comment (andschwa) [inline samples/data-sealing/common/CMakeLists.txt]: @anakrish I could use your help here. It's bizarre, but when trying to change this from OBJECT to STATIC, the enclaves can't link correctly. It's been suggested by @CodeMonkeyLeet and @EMumau that maybe there's a circular dependency, but I'm not sure why that would be.

**Comment by andschwa** [comment]
Testing this requires building and installing OE somewhere with my patch in #1513 (also included here), and then building the sample separately against that installed OE. Something like:

```
cd ~/openenclave
mkdir build && cd build && cmake .. -DUSE_LIBSGX=ON -DCMAKE_INSTALL_PREFIX=~/openenclave-install && make install
cd ~/openenclave/samples/data-sealing
mkdir build && cd build && cmake .. -DCMAKE_PREFIX_PATH="~/openenclave-install" && make
```

File: samples/data-sealing/common/CMakeLists.txt (hunk 9 above)
Comment (anakrish) [reply samples/data-sealing/common/CMakeLists.txt]: Does adding STATIC still preserve the -fPIC compiler flag?

Comment (andschwa) [reply samples/data-sealing/common/CMakeLists.txt]: Not sure if it did or not, but I forced it to and still the same issue:

```
> ninja -v -j1
[1/15] cd /home/andschwa/openenclave/samples/data-sealing/build/common && /home/andschwa/openenclave-install/bin/oeedger8r --trusted /home/andschwa/openenclave/samples/data-sealing/datasealing.edl
Generating edge routines for the Open Enclave SDK.
Success.
[2/15] /usr/bin/clang++-7  -DOE_BUILD_ENCLAVE -DOE_USE_LIBSGX -I../ -Icommon -isystem /home/andschwa/openenclave-install/include -isystem /home/andschwa/openenclave-install/include/openenclave/3rdparty -isystem /home/andschwa/openenclave-install/include/openenclave/3rdparty/libcxx -isystem /home/andschwa/openenclave-install/include/openenclave/3rdparty/libc -fPIC   -m64 -fPIC -nostdinc -fno-stack-protector -fvisibility=hidden -nostdinc++ -std=gnu++11 -MD -MT common/CMakeFiles/common.dir/dispatcher.cpp.o -MF common/CMakeFiles/common.dir/dispatcher.cpp.o.d -o common/CMakeFiles/common.dir/dispatcher.cpp.o -c ../common/dispatcher.cpp
[3/15] /usr/bin/clang++-7  -DOE_BUILD_ENCLAVE -DOE_USE_LIBSGX -I../ -Icommon -isystem /home/andschwa/openenclave-install/include -isystem /home/andschwa/openenclave-install/include/openenclave/3rdparty -isystem /home/andschwa/openenclave-install/include/openenclave/3rdparty/libcxx -isystem /home/andschwa/openenclave-install/include/openenclave/3rdparty/libc -fPIC   -m64 -fPIC -nostdinc -fno-stack-protector -fvisibility=hidden -nostdinc++ -std=gnu++11 -MD -MT common/CMakeFiles/common.dir/keys.cpp.o -MF common/CMakeFiles/common.dir/keys.cpp.o.d -o common/CMakeFiles/common.dir/keys.cpp.o -c ../common/keys.cpp
[4/15] /usr/bin/clang-7 -DOE_BUILD_ENCLAVE -DOE_USE_LIBSGX -I../ -Icommon -isystem /home/andschwa/openenclave-install/include -isystem /home/andschwa/openenclave-install/include/openenclave/3rdparty -isystem /home/andschwa/openenclave-install/include/openenclave/3rdparty/libc -fPIC   -m64 -fPIC -nostdinc -fno-stack-protector -fvisibility=hidden -MD -MT common/CMakeFiles/common.dir/datasealing_t.c.o -MF common/CMakeFiles/common.dir/datasealing_t.c.o.d -o common/CMakeFiles/common.dir/datasealing_t.c.o   -c common/datasealing_t.c
[5/15] : && /usr/local/cmake-3.13.1-Linux-x86_64/bin/cmake -E remove common/libcommon.a && /usr/bin/ar qc common/libcommon.a  common/CMakeFiles/common.dir/dispatcher.cpp.o common/CMakeFiles/common.dir/keys.cpp.o common/CMakeFiles/common.dir/datasealing_t.c.o && /usr/bin/ranlib common/libcommon.a && :
[6/15] /usr/bin/clang++-7  -DOE_BUILD_ENCLAVE -DOE_USE_LIBSGX -I../ -Icommon -isystem /home/andschwa/openenclave-install/include -isystem /home/andschwa/openenclave-install/include/openenclave/3rdparty -isystem /home/andschwa/openenclave-install/include/openenclave/3rdparty/libcxx -isystem /home/andschwa/openenclave-install/include/openenclave/3rdparty/libc -m64 -fPIC -nostdinc -fno-stack-protector -fvisibility=hidden -nostdinc++ -MD -MT enc1/CMakeFiles/enclave_a_v1.dir/ecalls.cpp.o -MF enc1/CMakeFiles/enclave_a_v1.dir/ecalls.cpp.o.d -o enc1/CMakeFiles/enclave_a_v1.dir/ecalls.cpp.o -c ../enc1/ecalls.cpp
[7/15] : && /usr/bin/clang++-7     enc1/CMakeFiles/enclave_a_v1.dir/ecalls.cpp.o  -o enc1/enclave_a_v1  common/libcommon.a /home/andschwa/openenclave-install/lib/openenclave/enclave/liboeenclave.a /home/andschwa/openenclave-install/lib/openenclave/enclave/libmbedtls.a /home/andschwa/openenclave-install/lib/openenclave/enclave/libmbedx509.a /home/andschwa/openenclave-install/lib/openenclave/enclave/libmbedcrypto.a /home/andschwa/openenclave-install/lib/openenclave/enclave/liboelibcxx.a /home/andschwa/openenclave-install/lib/openenclave/enclave/liboelibc.a /home/andschwa/openenclave-install/lib/openenclave/enclave/liboecore.a -nostdlib -nodefaultlibs -nostartfiles -Wl,--no-undefined,-Bstatic,-Bsymbolic,--export-dynamic,-pie,--build-id && :
FAILED: enc1/enclave_a_v1 
: && /usr/bin/clang++-7     enc1/CMakeFiles/enclave_a_v1.dir/ecalls.cpp.o  -o enc1/enclave_a_v1  common/libcommon.a /home/andschwa/openenclave-install/lib/openenclave/enclave/liboeenclave.a /home/andschwa/openenclave-install/lib/openenclave/enclave/libmbedtls.a /home/andschwa/openenclave-install/lib/openenclave/enclave/libmbedx509.a /home/andschwa/openenclave-install/lib/openenclave/enclave/libmbedcrypto.a /home/andschwa/openenclave-install/lib/openenclave/enclave/liboelibcxx.a /home/andschwa/openenclave-install/lib/openenclave/enclave/liboelibc.a /home/andschwa/openenclave-install/lib/openenclave/enclave/liboecore.a -nostdlib -nodefaultlibs -nostartfiles -Wl,--no-undefined,-Bstatic,-Bsymbolic,--export-dynamic,-pie,--build-id && :
/home/andschwa/openenclave-install/lib/openenclave/enclave/liboecore.a(calls.c.o): In function `_handle_call_enclave_function':
/home/andschwa/openenclave/build/package/../../enclave/core/sgx/calls.c:(.text+0x397d): undefined reference to `__oe_ecalls_table_size'
/home/andschwa/openenclave/build/package/../../enclave/core/sgx/calls.c:(.text+0x3ab6): undefined reference to `__oe_ecalls_table'
clang: error: linker command failed with exit code 1 (use -v to see invocation)
ninja: build stopped: subcommand failed.
```

Comment (anakrish) [reply samples/data-sealing/common/CMakeLists.txt]: The issue is due to the linker line.

All .o files specified as part of the linker line will be included; while only needed objects in specified libraries (libcommon.a) will be included. 

To make the link work, there are two options:
1. Use whole-archive to link libcommon.a. 

`-Wl,--whole-archive libcommon.a -Wl,--no-whole-archive`

This has the same effect as if all the object files contained in libcommon.a were explicitly specified in the command line.

2. Specify libcommon.a after liboecore.a in the linker line.

Comment (andschwa) [reply samples/data-sealing/common/CMakeLists.txt]: THANK YOU

File: samples/data-sealing/CMakeLists.txt (hunk 7 above)
Comment (andschwa) [reply samples/data-sealing/CMakeLists.txt]: See #1538.

File: samples/data-sealing/CMakeLists.txt (hunk 8 above)
Comment (andschwa) [reply samples/data-sealing/CMakeLists.txt]: Moved to #1539.

**Comment by andschwa** [comment]
bors try

**Comment by andschwa** [comment]
@johnkord the failures seem to provide an answer to the question I posed to you about why the samples are skipped in simulation mode 😆

**Comment by andschwa** [comment]
bors try

**Comment by andschwa** [comment]
Okay the samples should not properly be skipped in simulation mode.

bors try

File: samples/data-sealing/common/CMakeLists.txt [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+add_library(common STATIC dispatcher.cpp keys.cpp ${CMAKE_CURRENT_BINARY_DIR}/datasealing_t.c)
+target_link_libraries(common PUBLIC
+  # `liboecore`, a dependency of `liboeenclave`, requires the ecalls
+  # function table. Because the libraries linking `libcommon` do not
+  # directly require this symbol, the linker skips the object in
+  # `libcommon` which defines them. So we manually require it.
+  #
+  # Alternatively we could use a CMake OBJECT library, but that
+  # requires a newish version of CMake.
+  -Wl,--require-defined=__oe_ecalls_table
```
Comment (radhikaj) [inline samples/data-sealing/common/CMakeLists.txt]: Would everyone implementing a sample with cmake need to force the linker to include __oe_ecalls_table?

File: samples/data-sealing/datasealing.edl [hunk 11]
```diff
@@ -3,16 +3,16 @@
...
-    include "../shared.h"
+    include "common/shared.h"
     trusted {
-        public int seal_data(int sealPolicy, 
+        public int seal_data(int sealPolicy,
                              [in, size = opt_msg_len] unsigned char* opt_mgs,
                              size_t opt_msg_len,
                              [in, size = data_size] unsigned char* data,
                              size_t data_size,
                              [out] sealed_data_t** sealed_data,
```
Comment (radhikaj) [inline samples/data-sealing/datasealing.edl]: sealed_data_t should be annotated with size or should be user_check. @anakrish ?

**Comment by andschwa** [comment]
Updated with a relocatable CMake package to get rid of a hack I really didn't like. Unfortunately, the old script testing pkg-config fails on my machine, on master, but not in CI, so here goes:

bors try

File: samples/data-sealing/common/CMakeLists.txt (hunk 10 above)
Comment (andschwa) [reply samples/data-sealing/common/CMakeLists.txt]: No, this is only if you use a common static library like done here. Perhaps I could make the comment more clear?

File: samples/data-sealing/datasealing.edl (hunk 11 above)
Comment (andschwa) [reply samples/data-sealing/datasealing.edl]: That might be true but I didn't touch these except maybe whitespace 😄

**Comment by andschwa** [comment]
Old and new samples scripts passing in CI:

```
        Start 172: samples_pkg_config

172/173 Test #172: samples_pkg_config .......................................................................................................................................   Passed   11.35 sec

        Start 173: samples

173/173 Test #173: samples ..................................................................................................................................................   Passed    5.54 sec
```

**Comment by andschwa** [comment]
All righty, all I need left is reviews/approval 🙏

**Comment by andschwa** [comment]
This now also resolves #1538 and #1539.

File: samples/data-sealing/common/CMakeLists.txt (hunk 10 above)
Comment (radhikaj) [reply samples/data-sealing/common/CMakeLists.txt]: Sure. This is really complicated, though :(.

File: samples/data-sealing/datasealing.edl (hunk 11 above)
Comment (radhikaj) [reply samples/data-sealing/datasealing.edl]: OK, lets look at it another time?

**Comment by andschwa** [comment]
Pending non-code changes: I'd like to rewrite the readme that's included with the CMake package to explain how to use it, and that it's no longer experimental.

File: samples/data-sealing/CMakeLists.txt (hunk 3 above)
Comment (CodeMonkeyLeet) [reply samples/data-sealing/CMakeLists.txt]: Can we track this as an unresolved issue? I don't believe we should remove the make samples, but they should still be convergent in their design and we should feel free to restructure the build structure as appopriate (e.g. as has been done with the common code, but not the test signing of individual enclaves and the generation of test keys)

File: samples/test-samples.cmake [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+  # The prior common cannot succeed unless all commands before also
+  # succeeded, so testing only the result of this is sufficient.
+  #
+  # TODO: An unfortunate side-effect of this placement is that a
+  # failed sample will cause the rest of the samples to be skipped.
+  if (TEST_RESULT)
+    message(FATAL_ERROR "Samples test '${SAMPLE}' failed!")
+  endif ()
+
+  # TODO: Build the sample with GNU Make.
```
Comment (CodeMonkeyLeet) [inline samples/test-samples.cmake]: This introduces a test coverage regression. Can we please ensure that the existing test suite continues to run before merging to master?

File: samples/test-samples.cmake [hunk 13]
```diff
@@ -0,0 +1,55 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+# This script requires the variables SOURCE_DIR, BUILD_DIR, and
+# PREFIX_DIR to be defined:
+#
+#     cmake -DSOURCE_DIR=~/openenclave -DBUILD_DIR=~/openenclave/build -DPREFIX_DIR=/opt/openenclave -P ~/openenclave/samples/test-samples.cmake
```
Comment (CodeMonkeyLeet) [inline samples/test-samples.cmake]: The script doesn't assert that these variables are defined though? It seems like it would allow you to make to the current folder by default?

File: cmake/openenclave-config.cmake.in [hunk 14]
```diff
@@ -3,19 +3,27 @@
...
-set(OE_MAJOR "@PROJECT_VERSION_MAJOR@")
-set(OE_MINOR "@PROJECT_VERSION_MINOR@")
-set(OE_REVISION "@PROJECT_VERSION_PATCH@")
-set(OE_VERSION "@PROJECT_VERSION@")
+# Reference data.
 set(OE_BRANCH "@GIT_BRANCH@")
 set(OE_COMMIT "@GIT_COMMIT@")
 set(OE_BUILD_NUMBER "@BUILD_NUMBER@")
+
+set_and_check(OE_LIBDIR "@PACKAGE_CMAKE_INSTALL_LIBDIR@")
```
[the remark this answers is not in the document]
Comment (andschwa) [reply cmake/openenclave-config.cmake.in]: They are created by the package module [`configure_package_config_file`](https://cmake.org/cmake/help/latest/module/CMakePackageConfigHelpers.html) with this data below:

```
PATH_VARS CMAKE_INSTALL_LIBDIR CMAKE_INSTALL_BINDIR CMAKE_INSTALL_DATADIR CMAKE_INSTALL_INCLUDEDIR
```

This makes it relocatable 😄

[the remark this answers is not in the document]
Comment (andschwa) [reply cmake/openenclave-config.cmake.in]: Trust me @CodeMonkeyLeet you probably don't want to read those docs. The relevant excerpt is:

> The variables <var1> to <varN> given as PATH_VARS are the variables which contain install destinations. For each of them the macro will create a helper variable PACKAGE_<var...>. These helper variables must be used in the FooConfig.cmake.in file for setting the installed location. They are calculated by configure_package_config_file so that they are always relative to the installed location of the package

and the rest explains how to do it properly. It was funny that we were using this module, but not actually using it 😅

File: samples/CMakeLists.txt [hunk 15]
```diff
@@ -1,51 +1,23 @@
...
+install(FILES Makefile README.md
+  DESTINATION ${CMAKE_INSTALL_DATADIR}/openenclave/samples)
+
+# TODO: Remove this when the new script supports all the samples for
+# both the CMake package and pkg-config.
+add_test(samples_pkg_config
+  sh ${CMAKE_CURRENT_SOURCE_DIR}/test-samples.sh -i ${PROJECT_BINARY_DIR} ${CMAKE_INSTALL_PREFIX} ${PROJECT_BINARY_DIR}/samples/samples-test)
+set_property(TEST samples_pkg_config PROPERTY ENVIRONMENT "USE_LIBSGX=${USE_LIBSGX}")
+
+# This script currently only tests data-sealing against the CMake package.
```
[the remark this answers is not in the document]
Comment (andschwa) [reply samples/CMakeLists.txt]: I wanted to explain why there was a second test that seemingly does the same thing. I'll make sure it is deleted when the old script is deleted.

File: samples/data-sealing/CMakeLists.txt (hunk 3 above)
Comment (andschwa) [reply samples/data-sealing/CMakeLists.txt]: This actually worked out rather well, as we want a custom target to run the sample, since the actual test is located a context up in the 'upstream' project. So the only question is the name, and I see no compelling reason to change it. If this were standalone, then maybe `add_test(foo)` but it'd be more complicated, and not worth it IMO.

File: samples/data-sealing/README.md [hunk 16]
```diff
@@ -142,12 +140,9 @@ int unseal_data(sealed_data_t* sealed_data, size_t sealed_data_size,
...
 
-To build a sample, change directory to your target sample directory and run `make build` to build the sample and run `make run` to run it.
-
-For example:
-
 ```bash
-yourusername@yourVMname:~/openenclave/share/openenclave/samples$ cd data-sealing
-yourusername@yourVMname:~/openenclave/share/openenclave/samples/data-sealing$ make build
-yourusername@yourVMname:~/openenclave/share/openenclave/samples/data-sealing$ make run
+$ cd data-sealing
```
[the remark this answers is not in the document]
Comment (andschwa) [reply samples/data-sealing/README.md]: Oh, sure thing.

File: samples/data-sealing/common/CMakeLists.txt [hunk 17]
```diff
@@ -0,0 +1,22 @@
...
+# Create a library common to each of our three enclaves.
+add_library(common STATIC dispatcher.cpp keys.cpp ${CMAKE_CURRENT_BINARY_DIR}/datasealing_t.c)
+target_compile_definitions(common PUBLIC OE_API_VERSION=2)
+target_link_libraries(common PUBLIC
+  # `liboecore`, a dependency of `liboeenclave`, requires the ecalls
+  # function table. Because the libraries linking `libcommon` do not
+  # directly require this symbol, the linker skips the object in
+  # `libcommon` which defines them. So we manually require it.
+  #
+  # Alternatively we could use a CMake OBJECT library, but that
```
[the remark this answers is not in the document]
Comment (andschwa) [reply samples/data-sealing/common/CMakeLists.txt]: 3.12 is for the use of `OBJECT` libraries which we didn't require here. I am quite confident this will work with 3.10, and fairly confident it would work down to 3.x ish since it's a standard export. It would be nice to test that, but let's punt it for now?

File: samples/data-sealing/common/CMakeLists.txt (hunk 10 above)
Comment (andschwa) [reply samples/data-sealing/common/CMakeLists.txt]: I know! I told @soccerGB he gave me the worst sample. This exhibits the creation of a static library of common enclave code, linked into multiple enclaves, which is an edge case which requires this complicated-ness (though it's not that different from similar scenarios that aren't in enclaves) and is valuable to have as an example.

@aserdean is going to follow up and implement CMake package based builds for the rest of the samples, which should be much simpler and therefore easier to follow as well. But he's blocked until this goes in.

File: samples/data-sealing/common/shared.h (hunk 6 above)
Comment (andschwa) [reply samples/data-sealing/common/shared.h]: Ah, @anakrish I knew there was a header ordering issue. Still, IWYU.

File: samples/data-sealing/enc1/ecalls.cpp [hunk 18]
```diff
@@ -1,10 +1,10 @@
 // Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License.
 
+#include <common/datasealing_t.h>
```
[the remark this answers is not in the document]
Comment (andschwa) [reply samples/data-sealing/enc1/ecalls.cpp]: Yes, the discrepancy is caused by using oeedgr8r correctly in the CMake sample: the headers `datasealing_t.h` and `datasealing_u.h` are generated in the associated _build_ directory for `libcommon`, whereas the GNU Make build generates them in place (polluting the source tree). So with `libcommon` being a separate library, it made more sense (to me) to turn these into `<common/*>` headers, where `libcommon` has the correct pair of folders in its `INTERFACE_INCLUDE_DIRECTORIES`. This means we don't have to specify a build directory (which could be _anywhere_), and the headers are grouped by their library `libcommon`, and the _only_ change necessary to the GNU Makefiles is the use of `-I..`

In short, it was the simplest solution I came up with. If you want to provide another one, I'm all ears 😄 but it is more complicated than it looks!

File: samples/test-samples.cmake (hunk 13 above)
Comment (andschwa) [reply samples/test-samples.cmake]: This script isn't meant to be used outside of our test; I don't see a reason to add a nice UI to it.

File: samples/test-samples.cmake (hunk 12 above)
Comment (andschwa) [reply samples/test-samples.cmake]: This is not quite correct. The existing tests are still covered via the renamed test `samples_pkg_config`. The TODO here is staging for @aserdean to come in and finish #1357 by adding pkg-config sample test support here. But there is no test regression in this PR.

File: samples/CMakeLists.txt (hunk 15 above)
[the remark this answers is not in the document]
Comment (andschwa) [reply samples/CMakeLists.txt]: For reference, this is why there is no test regression below, only the addition of a test for the CMake package sample build.

File: samples/data-sealing/README.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (andschwa) [reply samples/data-sealing/README.md]: I also improved the readme a bit here.

File: samples/test-samples.cmake (hunk 12 above)
Comment (CodeMonkeyLeet) [reply samples/test-samples.cmake]: Got it, thanks for clarifying.

File: samples/data-sealing/enc1/ecalls.cpp (hunk 18 above)
[the remark this answers is not in the document]
Comment (CodeMonkeyLeet) [reply samples/data-sealing/enc1/ecalls.cpp]: I'm more concerned that the nuances of this are documented somewhere for maintainers so we can scale this to more sample authors. If you could capture this somewhere for reference, it would be helpful.

File: samples/data-sealing/common/CMakeLists.txt (hunk 17 above)
[the remark this answers is not in the document]
Comment (CodeMonkeyLeet) [reply samples/data-sealing/common/CMakeLists.txt]: I'm fine with punting, but as always, please track with an issue. We want to get to a point where we can be very crisp on the level of tooling support for SDK package users, and if that's different from maintainers.

File: samples/test-samples.cmake (hunk 13 above)
Comment (CodeMonkeyLeet) [reply samples/test-samples.cmake]: It's fine, but the observation is less about nice UI and more about defensive coding; it's parameterized rather than hardcoded i.e. there's an assumption of flexible use beyond it's invocation from samples/CMakeLists.txt, otherwise, it would be more direct to assign `${CMAKE_CURRENT_SOURCE_DIR}` internal to this script and less prone to usage error.

**Comment by soccerGB** [comment]
Changes look pretty clean.

I have two suggestions:

1.  I would be great if the -DCAMKE_PREFIX is not needed during the project building, which could be 
    achieved by adding " export OpenEnclave_DIR=path to the install package" into openenclaverc file for setting up package searching directory. I think this will be a better dev experience

2. The old Makefile lives in the same directory as the CMakeList.txt, which will get overwirtten after building with cmake.

File: samples/test-samples.cmake (hunk 13 above)
Comment (andschwa) [reply samples/test-samples.cmake]: It can't be hard-coded because the input directories change based on the location of the build (which includes absolute directories, so it changes with the environment). The upstream build (that is, Open Enclave) has this knowledge, but the downstream build doesn't have access to it, and they can't be relative paths (since you could put a build folder anywhere; in fact, I frequently nest them). I admit it's hacky, but we have to pass that info to it somehow. I wouldn't say the script is flexible in use beyond this scenario (at least, not without more work).

**Comment by andschwa** [comment]
@soccerGB 

For #1, that would be quite reasonable, can you open a PR after this goes in to update `openenclaverc` (and the readme I just added)?

For #2, this will only happen if you build in-source, which is not recommended (nor tested nor supported). If you `mkdir build && cd build && cd cmake ..` the existing GNU Makefiles will not be touched.

**Comment by soccerGB** [comment]
For #1,  I think it would cleaner to include the necessary changes to this PR instead of a separate one. 

I agree with your comment on  #2, I am fine with current approach.

File: samples/test-samples.cmake [hunk 19]
```diff
@@ -0,0 +1,55 @@
...
+
+# The samples cannot run in simulation mode.
```
Comment (aserdean) [inline samples/test-samples.cmake]: FWIW: Some samples can be ran in simulation mode (i.e. https://github.com/Microsoft/openenclave/blob/master/samples/helloworld/Makefile#L16)

Comment (andschwa) [reply samples/test-samples.cmake]: Ah, interesting. Thank you; I will let you separate them out and get what does work in simulation working 😄

File: samples/data-sealing/common/CMakeLists.txt (hunk 17 above)
[the remark this answers is not in the document]
Comment (andschwa) [reply samples/data-sealing/common/CMakeLists.txt]: Got it #1561.

File: samples/data-sealing/enc1/ecalls.cpp (hunk 18 above)
[the remark this answers is not in the document]
Comment (andschwa) [reply samples/data-sealing/enc1/ecalls.cpp]: Most of this can be avoided with better GNU Make builds for the samples, which @EMumau was going to take a look at. I'll file an issue.

[the remark this answers is not in the document]
Comment (andschwa) [reply samples/data-sealing/enc1/ecalls.cpp]: #1562
